src/rate_limiter.cc

```diff
@@ -538,24 +568,39 @@ RateLimiter::ModelContext::AllocateInstanceIfAvailable()
 }
 
 void
-RateLimiter::ModelContext::AddSpecificRequestQueue()
+RateLimiter::ModelContext::AddSpecificRequestQueue(
+    ModelInstanceContext* instance)
 {
   std::lock_guard<std::recursive_mutex> lk(sched_request_queue_mtx_);
-  specific_sched_request_queues_.emplace_back();
+  specific_sched_request_queues_[instance->RawInstance()];
 }
 
 bool
-RateLimiter::ModelContext::ContainsPendingRequests(int index)
+RateLimiter::ModelContext::ContainsPendingRequests(
+    ModelInstanceContext* instance)
 {
   std::lock_guard<std::recursive_mutex> lk(sched_request_queue_mtx_);
   return (generic_sched_request_queue_.size() != 0) ||
-         (specific_sched_request_queues_[index].size() != 0);
+         (specific_sched_request_queues_[instance->RawInstance()].size() != 0);
 }
 
 void
-RateLimiter::ModelContext::RequestRemoval()
+RateLimiter::ModelContext::RemoveInstance(ModelInstanceContext* instance)
 {
-  removal_in_progress_ = true;
+  std::lock_guard<std::recursive_mutex> lk1(sched_request_queue_mtx_);
+  std::lock_guard<std::recursive_mutex> lk2(avbl_instances_mtx_);
+
+  PriorityQueue new_avbl_instances;
+  while (!avbl_instances_.empty()) {
```

Comment:
What if the instance is not in the avbl_instance_ when the RemoveInstance is invoked?

Reply:
Update commit: [Use function instead of macro for ShareBackendThread](https://github.com/triton-inference-server/core/pull/189/commits/16d0054d7f5c7e63da636d180a1eeb4af61e1ae9)

Reply:
I think inference will stop and remain stopped on the removing instance before it is removed from the rate limiter, but I will verify it with the [new rate limiter test](https://github.com/triton-inference-server/server/pull/5635#issuecomment-1512273360) that will vary the resource while adding/removing instances.